skbio/sequence/_substitution.py

```diff
@@ -0,0 +1,518 @@
+        This is an alias of ``ids``.
+
+        """
+        return self._ids
+
+    @property
+    @experimental(as_of='0.5.10')
+    def scores(self):
+        """Matrix of substitution scores.
+
+        Each value corresponds to the score of substituting the row character
+        with the column character.
+
+        Returns
+        -------
+        2D np.ndarray
+            Matrix of substitution scores.
+
+        Notes
+        -----
+        This is an alias of ``data``.
+
+        """
+        return self._data
+
+    @experimental(as_of='0.5.10')
+    def __init__(self, alphabet, scores, **kwargs):
+        """Initialize a substitution matrix object with an alphabet and a score
+        matrix.
+        """
+        super().__init__(scores, alphabet, **kwargs)
+
+    @experimental(as_of='0.5.10')
+    def to_dict(self):
+        """Create a 2D dictionary from the substitution matrix.
+
+        Returns
+        -------
+        dict of dict
+            2D dictionary constructed from the substitution matrix.
+
+        """
+        return {id_: dict(zip(self._ids, row)) for id_, row in zip(
+            self._ids, self._data)}
+
+    @classonlymethod
+    @experimental(as_of='0.5.10')
+    def from_dict(cls, dictionary):
+        """Create a substitution matrix from a 2D dictionary.
+
+        Parameters
+        ----------
+        dictionary : dict of dict
+            2D dictionary of substitution scores from outer characters to inner
+            characters.
+
+        Returns
+        -------
+        SubstitutionMatrix
+            Substitution matrix constructed from the dictionary.
+
+        Raises
+        ------
+        ValueError
+            If outer and inner characters are inconsistent.
+        ValueError
+            If scores are not numbers.
+
+        Examples
+        --------
+        >>> from skbio import SubstitutionMatrix
+        >>> d = {'a': {'a': 1, 'b': 0, 'c': 0},
+        ...      'b': {'a': 0, 'b': 1, 'c': 0},
+        ...      'c': {'a': 0, 'b': 0, 'c': 1}}
+        >>> mat = SubstitutionMatrix.from_dict(d)
+        >>> mat.alphabet
+        ('a', 'b', 'c')
+        >>> mat.scores
+        array([[ 1.,  0.,  0.],
+               [ 0.,  1.,  0.],
+               [ 0.,  0.,  1.]])
+
+        """
+        alphabet = tuple(dictionary.keys())
+        alphabet_set = set(alphabet)
+        idmap = {x: i for i, x in enumerate(alphabet)}
+        scores = np.zeros((n := len(alphabet), n))
+        for i, row in enumerate(dictionary.values()):
+            if set(row.keys()) != alphabet_set:
+                raise ValueError('The outer and inner layers of the dictionary'
+                                 ' must have the same set of keys.')
+            for key, value in row.items():
+                if isinstance(value, int):
+                    value = float(value)
+                elif not isinstance(value, float):
+                    raise ValueError('Scores must be integers or floating-'
```

Comment:
I don't think this check or the preceding is necessary:

```python
In [5]: skbio.DistanceMatrix(np.array([['A','B'],['B','A']]), ['foo', 'bar'])
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Cell In[5], line 1
----> 1 skbio.DistanceMatrix(np.array([['A','B'],['B','A']]), ['foo', 'bar'])

File ~/miniconda3/envs/qiime2-amplicon-2023.9/lib/python3.8/site-packages/skbio/stats/distance/_base.py:136, in DissimilarityMatrix.__init__(self, data, ids, validate)
    133         _issue_copy = False
    135 if _issue_copy:
--> 136     data = np.asarray(data, dtype='float')
    138 if data.ndim == 1:
    139     # We can assume squareform will return a symmetric square matrix
    140     # so no need for full validation.
    141     # Still do basic checks (e.g. zero length)
    142     # and id validation
    143     data = squareform(data, force='tomatrix', checks=False)

ValueError: could not convert string to float: 'A'
```

Reply:
Good point. Let me modify.